ccl/backupccl: TestProtectedTimestampsFailDueToLimits failed

[(ccl/backupccl).TestProtectedTimestampsFailDueToLimits failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2851188&tab=buildLog) on [master@f676f07152b8d0ca203831bb9e75d9b7d3ed5b79](https://github.com/cockroachdb/cockroach/commits/f676f07152b8d0ca203831bb9e75d9b7d3ed5b79):

```
=== RUN   TestProtectedTimestampsFailDueToLimits
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestProtectedTimestampsFailDueToLimits401751820
    test_log_scope.go:74: use -show-logs to present logs inline
E210406 05:42:07.407953 434466 kv/kvserver/replica_proposal.go:299  [n1,s1,r33/1:/Table/3{7-8}] 1  could not run async checksum computation (ID = 5580af63-7cf6-48b7-b87c-6a67645042f0): node unavailable; try another peer
E210406 05:42:07.411733 435671 kv/kvserver/consistency_queue.go:191  [n1,consistencyChecker,s1,r33/1:/Table/3{7-8}] 2  result is ambiguous (server shutdown)
E210406 05:42:07.413106 435671 kv/kvserver/queue.go:1093  [n1,consistencyChecker,s1,r33/1:/Table/3{7-8}] 3  result is ambiguous (server shutdown)
    testcluster.go:135: condition failed to evaluate within 45s: unexpectedly found 1 active spans:
             0.000ms      0.000ms    === operation:job
        goroutine 434311 [running]:
        runtime/debug.Stack(0xc002ed33b0, 0x7ec93e0, 0xc0027414e0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x809fd00, 0xc001dabb00, 0xc002ed33b0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x85
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers(0xc000138480, 0x7f5b100, 0xc0001a4018)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:135 +0x31f
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start.func2()
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:335 +0x74
        github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc00a05a470)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:110 +0x35
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc008c9e240, 0x7f5b100, 0xc0001a4010)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x346
        github.com/cockroachdb/cockroach/pkg/ccl/backupccl.TestProtectedTimestampsFailDueToLimits(0xc001dabb00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backup_test.go:6426 +0x4f5
        testing.tRunner(0xc001dabb00, 0x6bf5470)
        	/usr/local/go/src/testing/testing.go:1123 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x5bc
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestProtectedTimestampsFailDueToLimits401751820
--- FAIL: TestProtectedTimestampsFailDueToLimits (46.40s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=4

```
make stressrace TESTS=TestProtectedTimestampsFailDueToLimits PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestProtectedTimestampsFailDueToLimits.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/backupccl).TestProtectedTimestampsFailDueToLimits failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2855048&tab=buildLog) on [master@651184b3cfbd7c6d336bb49e7152e3d48ea06416](https://github.com/cockroachdb/cockroach/commits/651184b3cfbd7c6d336bb49e7152e3d48ea06416):

```
=== RUN   TestProtectedTimestampsFailDueToLimits
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestProtectedTimestampsFailDueToLimits142880842
    test_log_scope.go:74: use -show-logs to present logs inline
    testcluster.go:135: condition failed to evaluate within 45s: unexpectedly found 1 active spans:
             0.000ms      0.000ms    === operation:job
        goroutine 443282 [running]:
        runtime/debug.Stack(0xc003d1b610, 0x52a5440, 0xc005205b80)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x5479e80, 0xc003bcac00, 0xc003d1b610)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x6d
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers(0xc00253e000, 0x53366e0, 0xc000074118)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:135 +0x23d
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start.func2()
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:335 +0x45
        github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc0012f7c70)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:110 +0x25
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0080de200, 0x53366e0, 0xc000074110)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x268
        github.com/cockroachdb/cockroach/pkg/ccl/backupccl.TestProtectedTimestampsFailDueToLimits(0xc003bcac00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backup_test.go:6426 +0x430
        testing.tRunner(0xc003bcac00, 0x4ceebd0)
        	/usr/local/go/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x2b3
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestProtectedTimestampsFailDueToLimits142880842
--- FAIL: TestProtectedTimestampsFailDueToLimits (45.26s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestProtectedTimestampsFailDueToLimits PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestProtectedTimestampsFailDueToLimits.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/backupccl/backup_planning.go
+++ b/pkg/ccl/backupccl/backup_planning.go
@@ -47,6 +47,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage/cloudimpl"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/interval"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
@@ -1290,23 +1291,35 @@ func backupPlanHook(
 		// the job record creation happens transactionally.
 		plannerTxn := p.ExtendedEvalContext().Txn
 
+		// Construct the job and commit the transaction. Perform this work in a
+		// closure to ensure that the job is cleaned up if an error occurs.
 		var sj *jobs.StartableJob
-		jobID := p.ExecCfg().JobRegistry.MakeJobID()
-		if err := p.ExecCfg().JobRegistry.CreateStartableJobWithTxn(ctx, &sj, jobID, plannerTxn, jr); err != nil {
-			return err
-		}
-		if err := doWriteBackupManifestCheckpoint(ctx, jobID); err != nil {
-			return err
-		}
-		if err := protectTimestampForBackup(ctx, p, plannerTxn, jobID, spans, startTime, endTime, backupDetails); err != nil {
-			return err
-		}
+		if err := func() (err error) {
+			defer func() {
+				if err == nil || sj == nil {
+					return
+				}
+				if cleanupErr := sj.CleanupOnRollback(ctx); cleanupErr != nil {
+					log.Errorf(ctx, "failed to cleanup job: %v", cleanupErr)
+				}
+			}()
+			jobID := p.ExecCfg().JobRegistry.MakeJobID()
+			if err := p.ExecCfg().JobRegistry.CreateStartableJobWithTxn(ctx, &sj, jobID, plannerTxn, jr); err != nil {
+				return err
+			}
+			if err := doWriteBackupManifestCheckpoint(ctx, jobID); err != nil {
+				return err
+			}
+			if err := protectTimestampForBackup(ctx, p, plannerTxn, jobID, spans, startTime, endTime, backupDetails); err != nil {
+				return err
+			}
 
-		// We commit the transaction here so that the job can be started. This
-		// is safe because we're in an implicit transaction. If we were in an
-		// explicit transaction the job would have to be run with the detached
-		// option and would have been handled above.
-		if err := plannerTxn.Commit(ctx); err != nil {
+			// We commit the transaction here so that the job can be started. This
+			// is safe because we're in an implicit transaction. If we were in an
+			// explicit transaction the job would have to be run with the detached
+			// option and would have been handled above.
+			return plannerTxn.Commit(ctx)
+		}(); err != nil {
 			return err
 		}
 

--- a/pkg/ccl/backupccl/restore_planning.go
+++ b/pkg/ccl/backupccl/restore_planning.go
@@ -1891,17 +1891,29 @@ func doRestorePlan(
 	// the job record creation happens transactionally.
 	plannerTxn := p.ExtendedEvalContext().Txn
 
+	// Construct the job and commit the transaction. Perform this work in a
+	// closure to ensure that the job is cleaned up if an error occurs.
 	var sj *jobs.StartableJob
-	jobID := p.ExecCfg().JobRegistry.MakeJobID()
-	if err := p.ExecCfg().JobRegistry.CreateStartableJobWithTxn(ctx, &sj, jobID, plannerTxn, jr); err != nil {
-		return err
-	}
+	if err := func() (err error) {
+		defer func() {
+			if err == nil || sj == nil {
+				return
+			}
+			if cleanupErr := sj.CleanupOnRollback(ctx); cleanupErr != nil {
+				log.Errorf(ctx, "failed to cleanup job: %v", cleanupErr)
+			}
+		}()
+		jobID := p.ExecCfg().JobRegistry.MakeJobID()
+		if err := p.ExecCfg().JobRegistry.CreateStartableJobWithTxn(ctx, &sj, jobID, plannerTxn, jr); err != nil {
+			return err
+		}
 
-	// We commit the transaction here so that the job can be started. This is
-	// safe because we're in an implicit transaction. If we were in an explicit
-	// transaction the job would have to be created with the detached option and
-	// would have been handled above.
-	if err := plannerTxn.Commit(ctx); err != nil {
+		// We commit the transaction here so that the job can be started. This is
+		// safe because we're in an implicit transaction. If we were in an explicit
+		// transaction the job would have to be created with the detached option and
+		// would have been handled above.
+		return plannerTxn.Commit(ctx)
+	}(); err != nil {
 		return err
 	}
 

--- a/pkg/ccl/importccl/import_stmt.go
+++ b/pkg/ccl/importccl/import_stmt.go
@@ -969,21 +969,33 @@ func importPlanHook(
 		// the job record creation happens transactionally.
 		plannerTxn := p.ExtendedEvalContext().Txn
 
+		// Construct the job and commit the transaction. Perform this work in a
+		// closure to ensure that the job is cleaned up if an error occurs.
 		var sj *jobs.StartableJob
-		jobID := p.ExecCfg().JobRegistry.MakeJobID()
-		if err := p.ExecCfg().JobRegistry.CreateStartableJobWithTxn(ctx, &sj, jobID, plannerTxn, jr); err != nil {
-			return err
-		}
+		if err := func() (err error) {
+			defer func() {
+				if err == nil || sj == nil {
+					return
+				}
+				if cleanupErr := sj.CleanupOnRollback(ctx); cleanupErr != nil {
+					log.Errorf(ctx, "failed to cleanup job: %v", cleanupErr)
+				}
+			}()
+			jobID := p.ExecCfg().JobRegistry.MakeJobID()
+			if err := p.ExecCfg().JobRegistry.CreateStartableJobWithTxn(ctx, &sj, jobID, plannerTxn, jr); err != nil {
+				return err
+			}
 
-		if err := protectTimestampForImport(ctx, p, plannerTxn, jobID, spansToProtect, walltime, importDetails); err != nil {
-			return err
-		}
+			if err := protectTimestampForImport(ctx, p, plannerTxn, jobID, spansToProtect, walltime, importDetails); err != nil {
+				return err
+			}
 
-		// We commit the transaction here so that the job can be started. This
-		// is safe because we're in an implicit transaction. If we were in an
-		// explicit transaction the job would have to be run with the detached
-		// option and would have been handled above.
-		if err := plannerTxn.Commit(ctx); err != nil {
+			// We commit the transaction here so that the job can be started. This
+			// is safe because we're in an implicit transaction. If we were in an
+			// explicit transaction the job would have to be run with the detached
+			// option and would have been handled above.
+			return plannerTxn.Commit(ctx)
+		}(); err != nil {
 			return err
 		}
 

